utils_time: Add common methods for timer device to avocado-vt

sitoliu opened the pull request:
Add timezone sync between host and guest, and clocksource
setting method in guest and host.

ID: 1445708
Signed-off-by: Sitong Liu <siliu@redhat.com>

sitoliu commented:
python ConfigTest.py --testcase=timedrift.with_pvclock.shared_ntp_date.ntp.with_reboot --guestname=Win7/RHEL.7.5

smp_4.1024m.repeat1.run_test.Host_RHEL.m7.u5.qcow2.virtio_blk.up.virtio_net.Win7.x86_64.sp1.myprovider.timedrift.with_pvclock.shared_ntp_date.ntp.with_reboot: PASS (748.36 s)

smp_4.1024m.repeat1.run_test.Host_RHEL.m7.u5.qcow2.virtio_blk.up.virtio_net.RHEL.7.5.x86_64.myprovider.timedrift.with_pvclock.shared_ntp_date.ntp.with_reboot: PASS (486.63 s)

vivianQizhu commented:
get_host_xxx should be able to combined together with get_guest_xxx, only distinguish by a parameter "session".

hereischen commented on virttest/utils_time.py:
```diff
@@ -0,0 +1,229 @@
+        timez_city = tz.groups()[0]
+        timez_code = tz.groups()[1]
+    else:
+        timez_city = ""
```
Better put these two lines, before `try`
`timez_city = ""`
 `timez_code = ""`
Then, you don't need this `else`.

vivianQizhu replied:
Agree with @hereischen

sitoliu replied:
Thanks! Update it.

hereischen commented on virttest/utils_time.py:
```diff
@@ -0,0 +1,229 @@
+        o = process.system_output(params["check_host_tz_cmd"], timeout=240,
+                                  shell=True)
+    except process.CmdError:
+        logging.warn("Can not get host's timezone")
```
with `timez_city = ""`
`timez_code = ""`declared before `try`, you can, `return` them when there is a CmdError.
Because you would not able to get the `tz` in `o`

vivianQizhu replied:
How about return False here? Cause you can't execute below code without the output.

hereischen commented on virttest/utils_time.py:
```diff
@@ -0,0 +1,229 @@
+    """
+    try:
+        cs = session.cmd_output("cat %s" % params["clocksource_dir"])
+        if cs in list_guest:
```
This func is not well defined, `if cs not in list_guest`, this func has no return value. 
Please reorganize it.

hereischen commented on virttest/utils_time.py:
```diff
@@ -0,0 +1,229 @@
+from virttest import error_context
+
+
+def get_host_timezone(params):
```
Having discussed with @luckyh , it's better not to take `params` as an argument.
You can just put the cmd in your func body, please try cover as many distributions as possible.

vivianQizhu commented on virttest/utils_time.py:
```diff
@@ -0,0 +1,229 @@
+
+
+@error_context.context_aware
+def align_guestz_to_hostz(params, vm, session, zonecode, zonecity):
```
Suggest function name like sync_guest_time_zone(), current name confused me, but it's up to you.

vivianQizhu replied:
Or, simply, sync_time_zone(), since the function itself only sync the operating system time zone with given parameter `zonecode` and `zoncity`, not really matter if they are from host and to guest.

vivianQizhu commented on virttest/utils_time.py:
```diff
@@ -0,0 +1,229 @@
+        guest_tz = re.match(r'^(?:\s+Time zone:\s)(\w+\/\w+|UTC)(?:\s\(\w+'
+                            r',\s)([+|-]\d{4})\)$', o_guest, re.M | re.I)
+        if guest_tz.groups()[1] == zonecode:
+            return (True, session)
```
Why return session? It does not changed from it is passed here.

sitoliu replied:
Thanks, timezone updating works only when guest is booting from qemu commands again, so we should return a new session.

vivianQizhu replied:
I don't think you need to return the session, you have the vm recreated, the user can get session of the vm object if they need, here this method have no responsibility to open up a session, suppose that user do not need a session after execute this method, it is easily to ignore the returned session open here, and forget close this session.

sitoliu replied:
Thanks. I return this session because I introduced one session in arguments, and if this session be closed in this method, the code depends on that session later would fail. I would try a local session in this method and close it before return.

vivianQizhu commented on virttest/utils_time.py:
```diff
@@ -0,0 +1,229 @@
+        status = session.cmd_status(params["set_tz_cmd"] % tz_value)
+
+    if not status:
+        session = vm.reboot(session=session, timeout=600)
```
If set time zone failed, reboot the guest and return True directly?

sitoliu replied:
Thanks, update it.

vivianQizhu commented on virttest/utils_time.py:
```diff
@@ -0,0 +1,229 @@
+            og = re.search(r'(?:\(UTC([+|-]\d{2}:\d{2})?)', line,
+                           re.M | re.I)
+            if og and og.group(1):
+                og_ = re.sub(r'(\d{2}):(\d{2})', r'\1\2', og.group(1))
```
@sitoliu I would suggest use more meaningful variable name to make the code more readable: `guest_tzs` -> `guest_time_zones`, `tz_list` -> `time_zone_list`, `o` -> `output`, `og` ?

sitoliu replied:
`og` is return by `re.search`, is a temp value. Usually `og.group(0) = '(UTC+xx:00' `, `og.group(0) = '+xx:00'`, or `og.group(0) = '(UTC'` and `og.group(1) = null`. When `og.group(1) ` is null, it is a UTC standard time, we should add timezone code "+0000" manually.

vivianQizhu commented on virttest/utils_time.py:
```diff
@@ -0,0 +1,229 @@
+        guest_tz = []
+        tz_list = session.cmd_output(params["list_tz_cmd"])
+        tz_list = tz_list.splitlines()
+        for line in tz_list:
```
How about define `guest_tz = []` here? It seems not usable out of the loop.

sitoliu replied:
The `guest_tz = []` in line76 is to define and initialize, in line101 it is to clean up for next read.

sitoliu commented:
Thanks for reviewing, this patch will be updated later.

vivianQizhu commented on virttest/utils_time.py:
```diff
@@ -0,0 +1,229 @@
+            guest_tz = []
+
+        for value in guest_tzs:
+            if value[0] == zonecode:
```
I am not sure, but I guess you don't actually need `guest_tzs[]`, why not return True directly when `guest_bz[0]==zonecode`?

sitoliu replied:
Thanks, there is a mistake. Modified it. We need `guest_tzs[]` to get the right timezone code for windows guest, because we can only get a timezone name from windows guest by the `check_tz_cmd`, and we want to compare the guest's timezone code with host's timezone code.

vivianQizhu commented on virttest/utils_time.py:
```diff
@@ -0,0 +1,229 @@
+        # get the timezone value for windows guest
+        tz_value = ""
+        for value in guest_tzs:
+            if zonecode == value[0]:
```
Same judgement with line 85? Looks you got returned and no chance go here?

sitoliu replied:
Modified it. Could you help review it again? Thanks!

balamuruhans commented on virttest/utils_time.py:
```diff
@@ -0,0 +1,243 @@
+from avocado.utils import process
+
+
+def get_host_timezone():
```
`get_timezone(session=None)` can be introduced to get either host or guest based on session.

sitoliu replied:
Thanks for review. I updated `get_guest_clocksource` and `get_host_clocksource` to `get_clocksource` in below. Here I just want to get the timezone of host, the result will be used in  `sync_timezone()`.

balamuruhans commented on virttest/utils_time.py:
```diff
@@ -0,0 +1,243 @@
+    if tz:
+        timez_city = tz.groups()[0]
+        timez_code = tz.groups()[1]
+    return (timez_city, timez_code)
```
return as dict with `{'city_timezone': timez_city, 'code_timezone': timez_code}`

balamuruhans commented on virttest/utils_time.py:
```diff
@@ -0,0 +1,243 @@
+    if os_type == "linux":
+        guest_tz = re.match(r'^(?:\s+Time zone:\s)(\w+\/\w+|UTC)(?:\s\(\w+'
+                            r',\s)([+|-]\d{4})\)$', o_guest, re.M | re.I)
+        if guest_tz.groups()[1] == zonecode:
```
some code redundancy can be avoided using 
```python
timezone = get_timezone(session)
if timezone['code_timezone'] == zonecode:
    return ...
else:
    status = timezone[city_timezone]
```

sitoliu replied:
Thanks! The patterns for timezone is different in linux and windows, `get_timezone()` only accomplish one pattern for linux, so I separate it here.

balamuruhans commented on virttest/utils_time.py:
```diff
@@ -0,0 +1,243 @@
+        return (False, session)
+
+
+def set_guest_clocksource(session, list_guest, expected):
```
`set_guest_clocksource`, `get_guest_clocksource`, `set_host_clocksource` and `get_host_clocksource` can be clubbed together as,
```python
def clock_source_info(list_info, expected=None, session=None, set=False):
    # For guest
    if session:
    # For host
    else:
```

balamuruhans commented on virttest/utils_time.py:
```diff
@@ -0,0 +1,243 @@
+        return None
+
+
+def get_guest_availble_clocksource(session, clocksource_list, dy_avail=False):
```
`get_guest_availble_clocksource` and `get_host_availble_clocksource` can be clubbed to `get_available_clock_source()`

balamuruhans commented:
More better way would be,

```python
class Timezone()
    def __init__():
        # Decide whether it is for guest/host
    def get_timezone():
    def set_timezone():
    def sync_timezone():
    def get_clock_source(available_source=False):
    def set_clock_source():
```

sitoliu commented:
Thanks for review, modified `get_timezone` method to get host or linux guest's timezone, and return a dict, and clubbed together the `get_clocksource` and `set_clocksource`.

balamuruhans commented on virttest/utils_time.py:
```diff
@@ -0,0 +1,208 @@
+    """
+    Handle host/guest's clocksource info.
+
+    session: VM session or host session.
```
:param session: VM session or host session

balamuruhans commented on virttest/utils_time.py:
```diff
@@ -0,0 +1,208 @@
+    Handle host/guest's clocksource info.
+
+    session: VM session or host session.
+    :param list_guest: Avaialble clocksource list for host.
```
not required, can be removed

balamuruhans commented on virttest/utils_time.py:
```diff
@@ -0,0 +1,208 @@
+                                                          clocksource_dir))
+        else:
+            status = process.system("echo %s > %s" % (expected,
+                                                      clocksource_dir))
```
`shell=True`

balamuruhans commented on virttest/utils_time.py:
```diff
@@ -0,0 +1,208 @@
+    Get current clocksource in use.
+
+    :session: VM session or host session.
+    :list_available: Available clocksource list.
```
```
:param list_available: Available clocksource list.
:param session: VM session or host session.
```

balamuruhans commented on virttest/utils_time.py:
```diff
@@ -0,0 +1,208 @@
+        return None
+
+
+def get_availble_clocksource(session=None, clocksource_list=None,
```
def get_available_clocksource

sitoliu commented:
@balamuruhans Has talked with @luckyh before, he recommends to use separated methods in utils_* file, like `get_timezone()`, which can be called directly by tp-qemu. If there is any update, please let me know, thanks!

balamuruhans commented:
> @balamuruhans Has talked with @luckyh before, he recommends to use separated methods in utils_* file, like get_timezone(), which can be called directly by tp-qemu. If there is any update, please let me know, thanks!
I am okay with implementation, code looks good to me. Thanks @sitoliu.

balamuruhans commented on virttest/utils_time.py:
```diff
@@ -0,0 +1,208 @@
+    try:
+        if session:
+            status = session.cmd_status(
+                    "echo %s > %s" % (expected, clocksource_dir), shell=True)
```
`TypeError: cmd_status() got an unexpected keyword argument 'shell'`

`shell=True ` we use it only for process

sitoliu replied:
Thanks so much! I will update it.

sitoliu commented:
Hi @hereischen , I am afraid I have to introduce params as an argument in `sync_timezone`, the vm should be destroyed and created again in windows guest after update timezone info. Do you have any suggestions? Thanks.

JOB LOG    : /root/avocado/job-results/job-2018-01-04T04.16-907ced8/job.log
 (1/2) smp_4.2048m.repeat1.pre_test.Host_RHEL.m7.u5.qcow2.virtio_scsi.up.virtio_net.Win2016.x86_64.myprovider.unattended_install.cdrom.extra_cdrom_ks.default_install.aio_threads: PASS (4741.22 s)
 (2/2) smp_4.2048m.repeat1.run_test.Host_RHEL.m7.u5.qcow2.virtio_scsi.up.virtio_net.Win2016.x86_64.myprovider.timedrift.with_pvclock.shared_ntp_date.ntp.with_reboot: PASS (789.28 s)
RESULTS    : PASS 2 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0

sitoliu commented:
Example of timezone list in windows guest.
`
(UTC+12:00) Fiji                 ------> get timezone code from this line.
Fiji Standard Time              ------> timezone name in windows guest.
....
(UTC) Coordinated Universal Time        ------> an utc timezone, add code info manually
UTC
....
(UTC-01:00) Azores
Azores Standard Time
.....
`

sitoliu commented:
Test PASS. 
(1/2) smp_8.8192m.repeat1.pre_test.Host_RHEL.m7.u5.qcow2.virtio_scsi.up.virtio_net.Win2016.x86_64.io-github-autotest-qemu.unattended_install.cdrom.extra_cdrom_ks.default_install.aio_native: PASS (1803.64 s)
 (2/2) smp_8.8192m.repeat1.run_test.Host_RHEL.m7.u5.qcow2.virtio_scsi.up.virtio_net.Win2016.x86_64.io-github-autotest-qemu.timedrift.with_pvclock.shared_ntp_date.ntp.with_reboot: PASS (793.81 s)

balamuruhans commented on virttest/utils_time.py:
```diff
@@ -0,0 +1,205 @@
+    try:
+        if session:
+            status = session.cmd_status(
+                    "echo %s > %s" % (expected, clocksource_dir))
```
form the string once and reuse it `cmd = "echo %s > %s" % (expected, clocksource_dir)` here and elsewhere. same in other methods.

balamuruhans commented on virttest/utils_time.py:
```diff
@@ -0,0 +1,205 @@
+        if not status:
+            return True
+        else:
+            return False
```
```python
if not status:
    return True
return False
```

sitoliu replied:
Updated, thanks.

sitoliu commented:
@vivianQizhu @hereischen @balamuruhans Update this patch, could you help review it again? Thanks so much.

balamuruhans commented on virttest/utils_time.py:
```diff
@@ -0,0 +1,244 @@
+    """
+    Update clocksource in kernel line with libguestfs.
+
+    :params: vm: Virtual machine for vm
```
param vm is not available for this method, it can be removed.

balamuruhans commented on virttest/utils_time.py:
```diff
@@ -0,0 +1,244 @@
+    :params: vm: Virtual machine for vm
+    :params: image_filename: Current image's file path.
+    :params: grub_file: VM's grub file path.
+    :params: kernel_cfg_pattern: Pattern to get current kernel line.
```
`%s/:params: kernel_cfg_pattern:/:params kernel_cfg_pattern:` please fix the same in elsewhere.

balamuruhans commented on virttest/utils_time.py:
```diff
@@ -0,0 +1,244 @@
+    command output in guest else return command ouput in host;
+
+    :param cmd: shell commands;
+    :param session: ShellSession or None;
```
1. param timeout is missed.
2. ; is not required.

hereischen commented on virttest/utils_time.py:
```diff
@@ -0,0 +1,244 @@
+    timezone_pattern = '^(?:\s+Time zone:\s)(\w+\/\w+|UTC)(?:\s\(\w+,\s)([+|-]\d{4})\)$'
+    error_context.context("Get host's timezone", logging.info)
+    host_timezone = process.system_output(timezone_cmd, timeout=240, shell=True)
+    host_timezone_set = re.match(timezone_pattern, host_timezone)
```
```
host_timezone_set = re.match(timezone_pattern, host_timezone).groups()
return {"timezone_city": host_timezone_set[0],
            "timezone_code": host_timezone_set[1]}
``` will save one function call

hereischen commented on virttest/utils_time.py:
```diff
@@ -0,0 +1,244 @@
+    timezone_cmd = 'timedatectl | grep "Time zone"'
+    timezone_pattern = '^(?:\s+Time zone:\s)(\w+\/\w+|UTC)(?:\s\(\w+,\s)([+|-]\d{4})\)$'
+    guest_timezone = session.cmd_output(timezone_cmd, timeout=240)
+    guest_timezone_set = re.match(timezone_pattern, guest_timezone)
```
```python
guest_timezone_set = re.match(timezone_pattern, guest_timezone).groups()
try:
    return (guest_timezone_set[0] == get_host_timezone()['timezone_city'])
except IndexError:
    raise exceptions.TestError("Fail to get guest's timezone.")
```

hereischen commented on virttest/utils_time.py:
```diff
@@ -0,0 +1,248 @@
+        session.cmd(set_timezone_cmd % host_timezone_city)
+        if not verify_timezone(session):
+            raise exceptions.TestError("Fail to sync guest's timezone.")
+    return True
```
Under the current implementation, this function doesn't need a return value.

hereischen commented on virttest/utils_time.py:
```diff
@@ -0,0 +1,248 @@
+        if verify_timezone_win(session) is not True:
+            raise exceptions.TestError("Fail to sync guest's timezone.")
+    session.close()
+    return True
```
same as line57

hereischen commented on virttest/utils_time.py:
```diff
@@ -0,0 +1,248 @@
+            else:
+                logging.warn("Can not get timezone name correctly.")
+            timezone_sets.append(timezone_set)
+            timezone_set = []
```
what does `timezone_set` contain? It seems its length will always be 1. 
When do you want to `timezone_set` to be an empty list?
If you want to make sure `timezone_set = []` in the `for` loop, it seems you can remove line 68 and 92, put `timezone_set = []` after line 72.

sitoliu replied:
Thanks. `timezone_set` is a set(not the python set) of timezone, such as `["+1200", "Fiji Standard Time"]`. It can get second value when program goes to `continue` in line80 or line84, and a new line will be in `line`. 
The format in `timezone_list` is:

>(UTC+12:00) Fiji
Fiji Standard Time
(empty line)
(UTC) Coordinated Universal Time
UTC

And `timezone_sets` is a list(python list) contains all the sets of timezone, like `[["+1200", "Fiji Standard Time"], ["-0100", "Azores Standard Time"],...]`.

vivianQizhu commented on virttest/utils_time.py:
```diff
@@ -0,0 +1,248 @@
+        raise exceptions.TestError("Fail to get guest's timezone.")
+
+
+def sync_timezone(session):
```
If it only works for linux, would it be better to name it like `sync_timezone_linux()`? the same for `verify_timezone()`

sitoliu replied:
Thanks, updated it.

sitoliu commented:
Test PASS in new installed RHEL.7.5 and Win2016.
1-smp_4.2048m.repeat1.pre_test.Host_RHEL.m7.u5.qcow2.virtio_blk.up.virtio_net.Guest.RHEL.7.5.x86_64.myprovider.unattended_install.cdrom.extra_cdrom_ks.default_install.aio_threads' Finshed 2018-03-14 02:56:42 2018-03-14 03:39:43 PASS 2580.453043
'2-smp_4.2048m.repeat1.pre_test.Host_RHEL.m7.u5.qcow2.virtio_blk.up.virtio_net.Guest.RHEL.7.5.x86_64.myprovider.rh_kernel_update' Finshed 2018-03-14 03:39:43 2018-03-14 03:43:39 PASS 235.772856
'3-smp_4.2048m.repeat1.run_test.Host_RHEL.m7.u5.qcow2.virtio_blk.up.virtio_net.Guest.RHEL.7.5.x86_64.myprovider.timedrift.shared_ntp_date.ntp.with_reboot' Finshed 2018-03-14 03:43:39 2018-03-14 03:45:56 PASS 137.284694

smp_4.2048m.repeat1.run_test.Host_RHEL.m7.u5.qcow2.virtio_blk.up.virtio_net.Guest.Win2016.x86_64.myprovider.timedrift.shared_ntp_date.ntp.with_reboot PASS.

hereischen commented on virttest/utils_time.py:
```diff
@@ -0,0 +1,258 @@
+                         " '%s' only" % (clksrc or "kvm-clock"))
+            kernel_cfg = re.findall(kernel_cfg_pattern,
+                                    kernel_cfg_original)[0]
+        except IndexError, detail:
```
should be `except IndexError as detail:`

hereischen commented on virttest/utils_time.py:
```diff
@@ -0,0 +1,258 @@
+        host_timezone_set = re.match(timezone_pattern, host_timezone).groups()
+        return {"timezone_city": host_timezone_set[0],
+                "timezone_code": host_timezone_set[1]}
+    except AttributeError, IndexError:
```
has to be `except (AttributeError, IndexError):`, and the rest of this kind of usage.

sitoliu replied:
Thanks, modified it.

hereischen commented:
Acked-by: Haotong Chen hachen@redhat.com

Please leave some comments in your code next time for the functions like `update_clksrc_with_kernel_line`.

hereischen commented on virttest/utils_time.py:
```diff
@@ -0,0 +1,258 @@
+
+
+@error_context.context_aware
+def update_clksrc_with_kernel_line(image_filename, grub_file,
```
BTW, I didn't see this function been used in your patches

vivianQizhu commented:
Test passed:
 (1/4) Host_RHEL.m7.u6.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.6.x86_64.myprovider.timedrift.shared_ntp_date.ntp.with_reboot: PASS (116.57 s)
 (2/4) Host_RHEL.m7.u6.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.6.x86_64.myprovider.timedrift.shared_ntp_date.date.with_reboot: PASS (86.05 s)
 (3/4) Host_RHEL.m7.u6.qcow2.virtio_scsi.up.virtio_net.Guest.Win10.x86_64.myprovider.timedrift.shared_ntp_date.ntp.with_reboot: PASS (176.85 s)
 (4/4) Host_RHEL.m7.u6.qcow2.virtio_scsi.up.virtio_net.Guest.Win10.x86_64.myprovider.timedrift.shared_ntp_date.date.with_reboot: PASS (236.32 s)

balamuruhans commented on virttest/utils_time.py:
```diff
@@ -0,0 +1,259 @@
+        vm.verify_alive()
+        session = vm.wait_for_login()
+        if verify_timezone_win(session) is not True:
+            raise exceptions.TestError("Fail to sync guest's timezone.")
```
It is better to close the `session` before raise the `exceptions`

balamuruhans commented on virttest/utils_time.py:
```diff
@@ -0,0 +1,259 @@
+    if session:
+        ret = session.cmd_output(cmd, timeout=timeout)
+    else:
+        ret = decode_to_text(process.system_output(cmd))
```
`cmd` from `sync_time_with_ntp()`requires `shell=True` so it is better to use `process.getoutput()` that takes care of it.

balamuruhans commented on virttest/utils_time.py:
```diff
@@ -0,0 +1,259 @@
+    cmd += "clocksource0/current_clocksource"
+    if expected not in execute(cmd, session=session):
+        return False
+    return True
```
```python
return expected in execute(cmd, session=session)
```

balamuruhans commented on virttest/utils_time.py:
```diff
@@ -0,0 +1,259 @@
+    """
+    error_context.context("Update guest kernel cli to '%s'" % (clksrc or
+                          "kvm-clock"), logging.info)
+    disk_obj = utils_disk.GuestFSModiDisk(image_filename)
```
is it mandatory to use libguestfs ? can we modify kernel commandline using API `update_boot_option` in utils_misc.py,

sitoliu commented:
@balamuruhans Thanks for review, updated my code accordingly and modified `update_clksrc` with using `update_boot_option`, will add the test result later.

sitoliu commented:
Result update, `update_clksrc` can work in the guest, while one case failed due to `tsc` is not the `available_clocksource` of rhel 7.6 guest, will fix it in tp-qemu case.
 
(1/4) Host_RHEL.m7.u6.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.6.x86_64.io-github-autotest-qemu.unattended_install.cdrom.extra_cdrom_ks.default_install.aio_threads: PASS (2849.29 s)
 (2/4) Host_RHEL.m7.u6.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.6.x86_64.io-github-autotest-qemu.rh_kernel_update: PASS (726.73 s)
 (3/4) Host_RHEL.m7.u6.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.6.x86_64.io-github-autotest-qemu.timerdevice.boot_test.with_reboot.reboot_immediately.without_host_load.clksource_kvm-clock.base_utc.clock_host: PASS (163.57 s)
 (4/4) Host_RHEL.m7.u6.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.6.x86_64.io-github-autotest-qemu.timerdevice.boot_test.with_reboot.reboot_immediately.without_host_load.clksource_tsc.base_utc.clock_host: ERROR (246.67 s)

@balamuruhans Thanks again for detailed advises, really help me! : - )

balamuruhans commented:
Thanks all, let us get this patch in.
